lib/aio/github.py

```diff
@@ -143,28 +143,25 @@
     def get_status(self, repo: str, sha: str, context: str | None, location: URL) -> Status:
         return GitHubStatus(self, repo, sha, context, location)
 
-    async def resolve_subject(
-        self, repo: str, sha: str | None, pull_nr: int | None, branch: str | None, target: str | None
-    ) -> Subject:
-        clone_url = self.clone / repo
+    async def resolve_subject(self, spec: SubjectSpecification) -> Subject:
+        clone_url = self.clone / spec.repo
 
-        if sha is not None:
+        if spec.sha is not None:
             # if pull_nr is set and our sha doesn't match the PR, we will
             # detect it soon
-            return Subject(clone_url, sha, target)
+            return Subject(clone_url, spec.sha, spec.target)
 
-        elif pull_nr is not None:
-            pull = await self.get_obj(f'repos/{repo}/pulls/{pull_nr}')
-            if target is None:
-                target = get_str(get_dict(pull, 'base'), 'ref')
+        elif spec.pull is not None:
+            pull = await self.get_obj(f'repos/{spec.repo}/pulls/{spec.pull}')
+            target = spec.target or get_str(get_dict(pull, 'base'), 'ref')
             return Subject(clone_url, get_str(get_dict(pull, 'head'), 'sha'), target)
 
         else:
-            if not branch:
-                branch = get_str(await self.get_obj(f'repos/{repo}'), 'default_branch')
+            if not spec.branch:
+                branch = get_str(await self.get_obj(f'repos/{spec.repo}'), 'default_branch')
 
-            with get_nested(await self.get_obj(f'repos/{repo}/git/refs/heads/{branch}'), 'object') as object:
-                return Subject(clone_url, get_str(object, 'sha'), target)
+            with get_nested(await self.get_obj(f'repos/{spec.repo}/git/refs/heads/{branch}'), 'object') as object:
```

Comment:
It's correct -- `branch` may be undefined here, if `spec.branch` is false-y. This was already wrong in the original code, but let's clean it up? Maybe `assert spec.branch` instead of `if`?

Reply:
... and your test caught it.  awesome 😎

Reply:
pyright also would have caught this, which is one of the big reasons I want to start using it.  mypy looks like they have some plans to support this, but nothing concrete yet: https://github.com/python/mypy/issues/2400

Reply:
> It's correct -- `branch` may be undefined here, if `spec.branch` is false-y. This was already wrong in the original code, but let's clean it up? Maybe `assert spec.branch` instead of `if`?

And for the record, this wasn't wrong in the original code.  It was a refactoring error — the original code had `branch` passed in as an argument, so it was always present, but usually `None`.